### PR TITLE
Changes to address issues #24 and #25

### PR DESCRIPTION
**Pull request by Andrer757**
[support] Updated dependencies versions.
[new] Added task to generate Version information class.
[new] Added task to update version on readme.
[new] Added options to show more logs on console and to redirect output to a file.
[new] Tasks now run per module.
[new] Added progress bar when output goes to file.
[new] Create and write to the output file when user add that option on console.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -11,12 +12,46 @@ repositories {
...
     compile group: 'commons-cli', name: 'commons-cli', version: '1.4'
-    compile 'com.google.code.gson:gson:2.8.0'
+    compile 'com.google.code.gson:gson:2.8.2'
+    compile 'me.tongfei:progressbar:0.5.5'
     runtime 'org.slf4j:slf4j-simple:1.7.10'
 }
 
+task generateVersion {
+    inputs.property "version", project.version
+    //outputs.dir "$buildDir/generated/source"
```
Comment (jzeferino) [inline build.gradle]: Remove this.

File: build.gradle [hunk 2]
```diff
@@ -11,12 +12,46 @@ repositories {
...
-    compile "org.gradle:gradle-tooling-api:4.0"
+    compile "org.gradle:gradle-tooling-api:4.5"
     compile group: 'commons-cli', name: 'commons-cli', version: '1.4'
-    compile 'com.google.code.gson:gson:2.8.0'
+    compile 'com.google.code.gson:gson:2.8.2'
+    compile 'me.tongfei:progressbar:0.5.5'
     runtime 'org.slf4j:slf4j-simple:1.7.10'
 }
 
+task generateVersion {
```
Comment (jzeferino) [inline build.gradle]: Just add some comment here explaining what this task do an whats for.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java [hunk 3]
```diff
@@ -20,56 +22,85 @@
...
      * @param args command line arguments.
      */
-    public static void main(String args[]) {
+    public static void main(String args[]) throws IOException {
 
         ProjectConnection gradleConnection = null;
+        ProgressBar pb = null;
+        String outputFile = null;
         try {
-            Pair<String, String> auth = CommandLineUtils.commandLineArgs(args);
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: If the `Pair ` class ins't used anymore, remove it.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java [hunk 4]
```diff
@@ -20,56 +22,85 @@
...
+            List<String> modules = configuration.getModules();
+
+            configuration.setBintrayUsername(auth.getUser());
+            configuration.setBintrayKey(auth.getKey());
+            configuration.setVerbose(auth.isVerbose());
+            outputFile = auth.getLogFile();
+            if (outputFile != null) {
+                File output = new File(outputFile);
+                if (output.exists()) {
+                    output.delete();
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: We should clear the file instead of delete it.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java [hunk 5]
```diff
@@ -20,56 +22,85 @@
...
 
             gradleConnection = GradleConnector.newConnector()
                     .forProjectDirectory(new File(configuration.getBasePath()))
                     .connect();
 
-            // Clean build and deploy all projects.
-            rebuildAndBintrayDeploy(gradleConnection, configuration);
+            ConsoleUtils.DrawInConsoleBox("Start process");
+
+            if (pb != null) pb.start();
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: Add {} here.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java [hunk 6]
```diff
@@ -20,56 +22,85 @@
...
+            //clean and build
+            for (String module : modules) {
+                if (pb != null) pb.setExtraMessage("Clean and building module " + module + "...");
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: the check for `pb != null` was made above. we should't need this here.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java [hunk 7]
```diff
@@ -20,56 +22,85 @@
...
+                GradleUtils.runGradle(gradleConnection, outputFile, new String[]{module + ":clean", module + ":build"}, configuration.getArguments());
+                if (pb != null) pb.step();
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: the check for pb != null was made above. we should't need this here.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java [hunk 8]
```diff
@@ -20,56 +22,85 @@
...
+            }
+
+            //clean and build
+            for (String module : modules) {
+                if (pb != null) pb.setExtraMessage("Uploading to bintray module " + module + "...");
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: the check for pb != null was made above. we should't need this here.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java [hunk 9]
```diff
@@ -20,56 +22,85 @@
...
+                GradleUtils.runGradle(gradleConnection, outputFile, new String[]{module + ":bintrayUpload"}, configuration.getArguments());
+                if (pb != null) pb.step();
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: the check for pb != null was made above. we should't need this here.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java [hunk 10]
```diff
@@ -20,56 +22,85 @@
...
+
         } catch (Exception e) {
-            System.out.println("Automator Error: " + e.toString());
+            if (pb != null) {
+                pb.stop();
+            }
+            if (outputFile == null) {
+                System.out.println("Automator Error: " + e.toString());
+            } else {
+                Writer output = new BufferedWriter(new FileWriter(outputFile));  //clears file every time
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: I don't understand this comment.

File: src/main/java/com/massivedisaster/bintraydeployautomator/model/Configuration.java [hunk 11]
```diff
@@ -100,6 +113,23 @@ public void setBintrayKey(String bintrayKey) {
         return extraTasks.toArray(new String[0]);
     }
 
+    private String[] getBuildTasks() {
+        if (modules == null) {
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/model/Configuration.java]: I don't really get why is this necessary.
modules should never been null. But if they are we aren't suppose to get here.

File: src/main/java/com/massivedisaster/bintraydeployautomator/utils/CommandLineUtils.java [hunk 12]
```diff
@@ -24,16 +32,18 @@
             cmd = parser.parse(options, args);
         } catch (ParseException e) {
             System.out.println(e.getMessage());
-            formatter.printHelp("Java -jar BintrayDeployAutomator-0.0.4.jar", options);
+            formatter.printHelp("Java -jar BintrayDeployAutomator-" + com.massivedisaster.bintraydeployautomator.utils.Version.VERSION + ".jar", options);
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/utils/CommandLineUtils.java]: `BintrayDeployAutomator-" + com.massivedisaster.bintraydeployautomator.utils.Version.VERSION + ".jar` is used in at least two palces. should be a final variable.

File: src/main/java/com/massivedisaster/bintraydeployautomator/utils/ConsoleUtils.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
+package com.massivedisaster.bintraydeployautomator.utils;
+
+public class ConsoleUtils {
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/utils/ConsoleUtils.java:3]: :heart: :dark_sunglasses:

File: src/main/java/com/massivedisaster/bintraydeployautomator/utils/GradleUtils.java [hunk 14]
```diff
@@ -15,12 +20,34 @@
...
+                    .forTasks(tasks)
+                    .withArguments(arguments)
+                    .setStandardOutput(os)
+                    .setStandardError(osErr)
+                    .run();
+
+            os.close();
+            osErr.close();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/utils/GradleUtils.java]: IMO, this should throw instead of catch and print.

File: src/main/java/com/massivedisaster/bintraydeployautomator/utils/GradleUtils.java [hunk 15]
```diff
@@ -15,12 +20,34 @@
...
-    public static void runGradle(@NotNull ProjectConnection gradleConnection, @NotNull String[] tasks, @NotNull String... arguments) {
-        gradleConnection.newBuild()
-                .forTasks(tasks)
-                .withArguments(arguments)
-                .setStandardOutput(System.out)
-                .setStandardError(System.err)
-                .run();
+    public static void runGradle(@NotNull ProjectConnection gradleConnection, @NotNull String output, @NotNull String[] tasks, @NotNull String... arguments) {
+        try {
+            OutputStream os;
```
Comment (jzeferino) [inline src/main/java/com/massivedisaster/bintraydeployautomator/utils/GradleUtils.java]: This file creation stuff should be in a separated method.

**Review by jzeferino — CHANGES_REQUESTED** [review]
:+1:

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java (hunk 3 above)
Comment (Andrer757) [reply src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: Pair class is from javafx.util.Pair not an actual class from the project, the import was removed to as you can see above.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java (hunk 6 above)
Comment (Andrer757) [reply src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: pb is only created when user adds the file to write the console output.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java (hunk 7 above)
Comment (Andrer757) [reply src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: pb is only created when user adds the file to write the console output.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java (hunk 8 above)
Comment (Andrer757) [reply src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: pb is only created when user adds the file to write the console output.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java (hunk 9 above)
Comment (Andrer757) [reply src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: pb is only created when user adds the file to write the console output.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java (hunk 10 above)
Comment (Andrer757) [reply src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: wrong copy/paste, fixed in next commit

File: src/main/java/com/massivedisaster/bintraydeployautomator/model/Configuration.java (hunk 11 above)
Comment (Andrer757) [reply src/main/java/com/massivedisaster/bintraydeployautomator/model/Configuration.java]: it can be null, we defined in the beginning that if there is no modules run the script for entire gradle project. Nonetheless this functions are no longer used so I will eliminate them in next commit.

File: src/main/java/com/massivedisaster/bintraydeployautomator/utils/CommandLineUtils.java (hunk 12 above)
Comment (Andrer757) [reply src/main/java/com/massivedisaster/bintraydeployautomator/utils/CommandLineUtils.java]: Fixed in the next commit.

File: src/main/java/com/massivedisaster/bintraydeployautomator/utils/GradleUtils.java (hunk 14 above)
Comment (Andrer757) [reply src/main/java/com/massivedisaster/bintraydeployautomator/utils/GradleUtils.java]: Fixed in next commit.

File: src/main/java/com/massivedisaster/bintraydeployautomator/utils/GradleUtils.java (hunk 15 above)
Comment (Andrer757) [reply src/main/java/com/massivedisaster/bintraydeployautomator/utils/GradleUtils.java]: Fixed in next commit.

File: build.gradle (hunk 2 above)
Comment (Andrer757) [reply build.gradle]: Done in next commit

File: build.gradle (hunk 1 above)
Comment (Andrer757) [reply build.gradle]: removed in next commit.

File: src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java (hunk 3 above)
Comment (jzeferino) [reply src/main/java/com/massivedisaster/bintraydeployautomator/Automator.java]: I had ideia that it was created by us.

**Review by jzeferino — APPROVED** [review]
This LGTM. Wonder if you want to improve the `if(pb!=null)` in this PR or you want to make a new one.

**Comment by Andrer757** [comment]
I will improve the if(pb!=null) in this PR.

**Review by jzeferino — APPROVED** [review]
LGTM now.
